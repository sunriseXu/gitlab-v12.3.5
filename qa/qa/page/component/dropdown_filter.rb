# frozen_string_literal: true

module QA
  module Page
    module Component
      module DropdownFilter
        def filter_and_select(item)
          wait(reload: false) do
            page.has_css?('.dropdown-input-field')
          end

          find('.dropdown-input-field').set(item)
          click_link item
        end
      end
    end
  end
end
