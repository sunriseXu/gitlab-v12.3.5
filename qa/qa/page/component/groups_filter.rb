# frozen_string_literal: true

module QA
  module Page
    module Component
      module GroupsFilter
        def self.included(base)
          base.view 'app/views/shared/groups/_search_form.html.haml' do
            element :groups_filter
          end

          base.view 'app/assets/javascripts/groups/components/groups.vue' do
            element :groups_list_tree_container
          end
        end

        private

        def has_filtered_group?(name)
          # Filter and submit to reload the page and only retrieve the filtered results
          find_element(:groups_filter).set(name).send_keys(:return)

          # Since we submitted after filtering, the presence of
          # groups_list_tree_container means we have the complete filtered list
          # of groups
          wait(reload: false) do
            page.has_css?(element_selector_css(:groups_list_tree_container))
          end

          # If there are no groups we'll know immediately because we filtered the list
          return false if page.has_text?('No groups or projects matched your search', wait: 0)

          # The name will be present as filter input so we check for a link, not text
          page.has_link?(name, wait: 0)
        end
      end
    end
  end
end
