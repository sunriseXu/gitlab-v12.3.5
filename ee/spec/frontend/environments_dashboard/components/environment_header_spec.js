import { shallowMount, createLocalVue } from '@vue/test-utils';
import { GlLink, GlBadge } from '@gitlab/ui';
import Icon from '~/vue_shared/components/icon.vue';
import ReviewAppLink from '~/vue_merge_request_widget/components/review_app_link.vue';
import component from 'ee/environments_dashboard/components/dashboard/environment_header.vue';

const localVue = createLocalVue();

describe('Environment Header', () => {
  const Component = localVue.extend(component);
  let wrapper;
  let propsData;

  beforeEach(() => {
    propsData = {
      environment: {
        environment_path: '/enivronment/1',
        name: 'staging',
        external_url: 'http://example.com',
        size: 1,
        within_folder: false,
      },
    };
  });

  afterEach(() => {
    wrapper.destroy();
  });

  describe('renders name and link to app', () => {
    beforeEach(() => {
      wrapper = shallowMount(Component, {
        propsData,
        localVue,
      });
    });

    it('renders the environment name', () => {
      expect(wrapper.find('.js-environment-name').text()).toBe(propsData.environment.name);
    });

    it('renders a link to the environment page', () => {
      expect(wrapper.find(GlLink).attributes('href')).toBe(propsData.environment.environment_path);
    });

    it('does not show a badge with the number of environments in the folder', () => {
      expect(wrapper.find(GlBadge).exists()).toBe(false);
    });

    it('renders a link to the external app', () => {
      expect(wrapper.find(ReviewAppLink).attributes('link')).toBe(
        propsData.environment.external_url,
      );
    });

    it('matches the snapshot', () => {
      expect(wrapper.element).toMatchSnapshot();
    });
  });

  describe('with environments grouped into a folder', () => {
    beforeEach(() => {
      propsData.environment.size = 5;
      propsData.environment.within_folder = true;
      propsData.environment.name = 'review/testing';

      wrapper = shallowMount(Component, {
        propsData,
        localVue,
      });
    });

    it('shows a badge with the number of other environments in the folder', () => {
      const expected = propsData.environment.size.toString();
      expect(wrapper.find(GlBadge).text()).toBe(expected);
    });

    it('shows an icon stating the environment is one of many in a folder', () => {
      expect(wrapper.find(Icon).attributes('name')).toBe('information');
      expect(wrapper.find(Icon).attributes('data-original-title')).toMatch(
        /last updated environment/,
      );
    });

    it('matches the snapshot', () => {
      expect(wrapper.element).toMatchSnapshot();
    });
  });

  describe('has errors', () => {
    beforeEach(() => {
      propsData.hasErrors = true;

      wrapper = shallowMount(Component, {
        propsData,
        localVue,
      });
    });

    it('matches the snapshot', () => {
      expect(wrapper.element).toMatchSnapshot();
    });
  });

  describe('has a failed pipeline', () => {
    beforeEach(() => {
      propsData.hasPipelineFailed = true;

      wrapper = shallowMount(Component, {
        propsData,
        localVue,
      });
    });

    it('matches the snapshot', () => {
      expect(wrapper.element).toMatchSnapshot();
    });
  });
});
