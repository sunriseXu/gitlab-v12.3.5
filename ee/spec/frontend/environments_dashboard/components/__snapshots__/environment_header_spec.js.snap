// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Environment Header has a failed pipeline matches the snapshot 1`] = `
<div
  class="card-header border-0 py-2 d-flex align-items-center dashboard-card-header-failed"
>
  <div
    class="flex-grow-1 block-truncated"
  >
    <gllink-stub
      class="js-environment-link cgray"
      data-original-title="staging"
      href="/enivronment/1"
      title=""
    >
      <span
        class="js-environment-name bold"
      >
         staging
      </span>
    </gllink-stub>
     
    <!---->
  </div>
   
  <reviewapplink-stub
    cssclass="btn btn-default btn-sm"
    link="http://example.com"
  />
</div>
`;

exports[`Environment Header has errors matches the snapshot 1`] = `
<div
  class="card-header border-0 py-2 d-flex align-items-center dashboard-card-header-warning"
>
  <div
    class="flex-grow-1 block-truncated"
  >
    <gllink-stub
      class="js-environment-link cgray"
      data-original-title="staging"
      href="/enivronment/1"
      title=""
    >
      <span
        class="js-environment-name bold"
      >
         staging
      </span>
    </gllink-stub>
     
    <!---->
  </div>
   
  <reviewapplink-stub
    cssclass="btn btn-default btn-sm"
    link="http://example.com"
  />
</div>
`;

exports[`Environment Header renders name and link to app matches the snapshot 1`] = `
<div
  class="card-header border-0 py-2 d-flex align-items-center bg-light"
>
  <div
    class="flex-grow-1 block-truncated"
  >
    <gllink-stub
      class="js-environment-link cgray"
      data-original-title="staging"
      href="/enivronment/1"
      title=""
    >
      <span
        class="js-environment-name bold"
      >
         staging
      </span>
    </gllink-stub>
     
    <!---->
  </div>
   
  <reviewapplink-stub
    cssclass="btn btn-default btn-sm"
    link="http://example.com"
  />
</div>
`;

exports[`Environment Header with environments grouped into a folder matches the snapshot 1`] = `
<div
  class="card-header border-0 py-2 d-flex align-items-center bg-light"
>
  <div
    class="flex-grow-1 block-truncated"
  >
    <gllink-stub
      class="js-environment-link cgray"
      data-original-title="review/testing"
      href="/enivronment/1"
      title=""
    >
      <span
        class="js-environment-name bold"
      >
         review/testing
      </span>
    </gllink-stub>
     
    <glbadge-stub
      class="dashboard-card-icon"
      pill="true"
    >
      5
    </glbadge-stub>
  </div>
   
  <icon-stub
    cssclasses="dashboard-card-icon"
    data-original-title="You are looking at the last updated environment"
    name="information"
    size="16"
    title=""
  />
</div>
`;
